egs/wsj/s5/steps/conf/train_calibration.sh

```diff
@@ -0,0 +1,120 @@
+# - log of frames per character ratio
+# (- categorical distribution of 'lang/words.txt', DISABLED)
+
+# begin configuration section.
+cmd=
+lmwt=12
+decode_mbr=true
+word_min_count=10 # Minimum word-count for single-word category,
+normalizer=0.0025 # L2 regularization constant,
+stage=0
+# end configuration section.
+
+[ -f ./path.sh ] && . ./path.sh
+. parse_options.sh || exit 1;
+
+if [ $# -ne 7 ]; then
+  echo "Usage: $0 [opts] <data-dir> <lang-dir|graph-dir> <word-filter> <word-length> <unigrams> <decode-dir> <calibration-dir>"
+  echo " Options:"
+  echo "    --cmd (run.pl|queue.pl...)      # specify how to run the sub-processes."
+  echo "    --lmwt <int>                    # scaling for confidence extraction"
+  echo "    --decode-mbr <bool>             # use Minimum Bayes Risk decoding"
+  echo "    --grep-filter <str>             # remove words from calibration targets"
+  exit 1;
+fi
+
+set -euxo pipefail
+
+data=$1
+lang=$2 # Note: may be graph directory not lang directory, but has the necessary stuff copied.
+word_filter=$3
+word_length=$4
+unigrams=$5
+latdir=$6
+dir=$7
+
+model=$latdir/../final.mdl # assume model one level up from decoding dir.
+
+for f in $data/text $lang/words.txt $word_filter $word_length $unigrams $latdir/lat.1.gz; do
+  [ ! -f $f ] && echo "$0: Missing file $f" && exit 1
+done
+[ -z "$cmd" ] && echo "$0: Missing --cmd '...'" && exit 1
+
+[ -d $dir/log ] || mkdir -p $dir/log
+nj=$(cat $latdir/num_jobs)
+
+# Store the setup,
+echo $lmwt >$dir/lmwt
+echo $decode_mbr >$dir/decode_mbr
+cp $word_filter $dir/word_filter
+cp $word_length $dir/word_length
+cp $unigrams $dir/unigrams
+
+# Create the ctm with raw confidences,
+# - we keep the timing relative to the utterance,
+if [ $stage -le 0 ]; then
+  $cmd JOB=1:$nj $dir/log/get_ctm.JOB.log \
+    lattice-scale --inv-acoustic-scale=$lmwt "ark:gunzip -c $latdir/lat.JOB.gz|" ark:- \| \
+    lattice-limit-depth ark:- ark:- \| \
+    lattice-push --push-strings=false ark:- ark:- \| \
+    lattice-align-words-lexicon --max-expand=10.0 \
+     $lang/phones/align_lexicon.int $model ark:- ark:- \| \
+    lattice-to-ctm-conf --decode-mbr=$decode_mbr ark:- - \| \
+    utils/int2sym.pl -f 5 $lang/words.txt \
+    '>' $dir/JOB.ctm
+  # Merge and clean,
+  set +x; for ((n=1; n<=nj; n++)); do cat $dir/${n}.ctm; done > $dir/ctm
+  rm $dir/*.ctm; set -x
+fi
+
+# Get evaluation of the 'ctm' using the 'text' reference,
+if [ $stage -le 1 ]; then
+  steps/conf/convert_ctm_to_tra.py $dir/ctm - | \
+  align-text --special-symbol="<eps>" ark:$data/text ark:- ark,t:- | \
+  utils/scoring/wer_per_utt_details.pl --special-symbol "<eps>" \
+  >$dir/align_text 
+  # Append alignment to ctm,
+  steps/conf/append_eval_to_ctm.py $dir/align_text $dir/ctm $dir/ctm_aligned
+fi
+
+# Prepare word-categories (based on wotd frequencies in 'ctm'),
+steps/conf/prepare_word_categories.py --min-count $word_min_count $lang/words.txt $dir/ctm $dir/word_categories
+
```

Comment:
Hi Dan, I waited some time to cool down and think it over. Okay, so how about creating a feature file for the words with the format:
word(str) id(int) filter(int) category(int) word-len(int) log-p-unigram(float) ... [extra feats]

The first 5 columns will have fixed order, while any number of columns can be added.
The first two columns are the 'words.txt'.

The filtering gets disabled by putting '1' everywhere. As it is now, the filtering means that words are excluded only from the training. The confidences get computed for all the words, including the partial words etc... I did check the NCE on 3 databases: switchboard, tedlium and ami/ihm. The filtering helped a little on switchboard, on other databases it had no effect. So I suppose it won't be harmful to other databases either.

The category will be encoded as an integer, 0 means the default category for infrequent an unseen words. The counts will be computed from file format same as 'text'. Ideally the categories should be computed from the 'ctm' that is used for calibration training, which does not exist in 'local/' before the 'steps/conf/train_calibration.sh' is called. So how about optionally passing the 'text' into the 'steps/conf/train_calibration.sh' to replace the 'ctm' by the 'train/text'? The alternative is to produce it outside, but then we will need to add dependency on the normal scoring (sometimes kaldi sometimes, sclite, probably you won't like this)...
So we'll better exclude that column from the 'external' word features.

Word lenght is the number of characters in the word.

And the log-p-unigram is the first column of the section that can have arbitrary number of features that get forwarded to the logistic regression input.

I don't plan to publish a paper about confidence calibration, the implementation follows standard techniques. Concerning the alternative way to compute the NCE, it might not be necessary at this moment. I'd like to finish this up to continue the work on the thesis. I primarily created the implementation for my personal use in the thesis, and as a courtesy I wanted to share it into kaldi. But I don't think I should spend too much time on it...